src/internet/radiobrowser/radiobrowserservice.cpp

```diff
@@ -0,0 +1,448 @@
+  ret.set_title(song_title);
+  ret.set_artist(name_);
+  ret.set_url(url_);
+  ret.set_art_automatic(favicon_.toString());
+  return ret;
+}
+
+void RadioBrowserService::Homepage() {
+  QDesktopServices::openUrl(homepage_url_);
+}
+
+PlaylistItem::Options RadioBrowserService::playlistitem_options() const {
+  return PlaylistItem::PauseDisabled | PlaylistItem::SeekDisabled;
+}
+
+void RadioBrowserService::PopulateCategory(QStandardItem* parentItem,
+                                           QStringList& elements) {
+  if (parentItem->hasChildren())
+    parentItem->removeRows(0, parentItem->rowCount());
+
+  for (const QString& element : elements) {
+    QStandardItem* item = new QStandardItem(
+        IconLoader::Load("icon_radio", IconLoader::Lastfm), QString());
+    item->setText(element);
+    item->setData(RadioBrowserService::Type_CategoryItem,
+                  InternetModel::Role_Type);
+    item->setData(true, InternetModel::Role_CanLazyLoad);
+    parentItem->appendRow(item);
+  }
+}
+
+void RadioBrowserService::PopulateStreams(QStandardItem* parentItem,
+                                          StreamList& streams) {
+  if (parentItem->hasChildren())
+    parentItem->removeRows(0, parentItem->rowCount());
+
+  for (const Stream& stream : streams) {
+    QStandardItem* item = new QStandardItem(
+        IconLoader::Load("icon_radio", IconLoader::Lastfm), QString());
+    item->setText(stream.name_);
+    item->setData(QVariant::fromValue(stream.ToSong(name_)),
+                  InternetModel::Role_SongMetadata);
+    item->setData(InternetModel::PlayBehaviour_SingleItem,
+                  InternetModel::Role_PlayBehaviour);
+    item->setData(InternetModel::Type_Track, InternetModel::Role_Type);
+
+    parentItem->appendRow(item);
+  }
+}
+
+void RadioBrowserService::ReloadSettings() {
+  QSettings s;
+  s.beginGroup(kSettingsGroup);
+
+  main_server_url_ =
+      s.value("server", RadioBrowserService::defaultServer).toString();
+
+  if (root_ && root_->hasChildren()) {
+    root_->removeRows(0, root_->rowCount());
+    root_->setData(true, InternetModel::Role_CanLazyLoad);
+  }
+}
+
+bool RadioBrowserService::EnsureServerConfig() {
+  if (main_server_url_.isEmpty())
+    return false;
+  else
+    return true;
+}
+
+void RadioBrowserService::LastRequestFailed() { ShowConfig(); }
+
+void RadioBrowserService::ShowConfig() {
+  app_->OpenSettingsDialogAtPage(SettingsDialog::Page_RadioBrowser);
+}
+
+void RadioBrowserService::Search(int search_id, const QString& query,
+                                 const int limit) {
+  QString determinedUrl =
+      RadioBrowserService::SearchUrl.arg(main_server_url_, query).arg(limit);
+  QUrl url(determinedUrl);
+
+  QNetworkReply* reply = network_->get(QNetworkRequest(url));
+  int task_id = app_->task_manager()->StartTask(tr("Getting channels"));
+
+  connect(reply, &QNetworkReply::finished,
+          [=] { this->SearchFinishedInternal(reply, task_id, search_id); });
```

Comment:
You could replace this with an `std::bind` but the lambda might be more readable.